lib/kiosk_example/udevd_server.ex

```
@@ -9,7 +9,7 @@ defmodule KioskExample.UdevdServer do
 
   @spec udevadm(String.t()) :: {Collectable.t(), exit_status :: non_neg_integer() | :timeout}
   def udevadm(args) do
-    MuonTrap.cmd("udevadm", String.split(args),
+    MuonTrap.cmd("udevadm", ~w"#{args}",
```

While I know this works for this instance, both `String.split/1` and `~w` ignore embedded double quotes in the command line. If this pattern is blindly copy/pasted to other places, it's going to introduce bugs. I think the `udevadm/1` function needs to a couple improvements:

1. Change `args` from a string to a list of strings. That will get rid of the need to `String.split/1` or `~w`.
2. Don't use `~w` to split the string when calling. That's the pattern that doesn't work in general for command line parameters. The call will be more verbose, but at least it will be correct if it gets copy/pasted to other code where arguments have spaces in them.
3. Check the result of the `MuonTrap.cmd/3` call in `udevadm` and log a message if it returns error. Given that errors are ignored in the calls, it would be good to at least see a message in the logs if the commands don't work. (Do they even work now?)

I think it will be quickest to split this commit out of the "Bump systems versions" PR since I think the systems update part of this is good.

---

@fhunleth Thank you for the review.

> Given that errors are ignored in the calls, it would be good to at least see a message in the logs if the commands don't work. (Do they even work now?)

Right. I currently checked it here, line 37, by `Process.flag(:trap_exit, true)`.

https://github.com/nerves-web-kiosk/kiosk_example/blob/ccdbd30253684d9815aae85e00f32047eaf6d9a5/lib/kiosk_example/udevd_server.ex#L31-L40

They puts the following logs. So they works even now.

```
21:47:20.270 [debug] udevadm (#Port<0.37>) exited by :normal.

21:47:20.270 [debug] udevadm (#Port<0.38>) exited by :normal.

21:47:20.270 [debug] udevadm (#Port<0.39>) exited by :normal.
```

I used the above codes for `udevd` but it is `start_link`ed, so I noticed that I don't need to use `Process.flag(:trap_exit, true)` .
So it can be simpler by not using it and checking return with  `{_, 0} = `.

---

I'm not particularly attached to w sigile. When quit to use it, I would like to change all of them. I also use the sigil several times for same pattern in this project.

First, let me clarify the point, I have a question.

Could you tell me the issue you mentioned, `both String.split/1 and ~w ignore embedded double quotes in the command line`. I still didn't get it. Did you mean like `"find . -name "*.ex""` things?

---

The example of where `String.split/1` won't work is for a command with an argument with a space in it. Like this:

```
rm "/data/my files/file.txt"
```

If you were to pass this to `sh`, `sh` would pass `rm` one argument with the contents `"/data/my file/file.txt"`. However, since this is being passed to `MuonTrap.cmd` (same with `System.cmd`), the arguments are passed as is. `String.split/1` would break the arguments into `"\"/data/my"` and `"files/file.txt"`, so `rm` would get two arguments and the `"` and wouldn't work.

Please feel free to use the w sigil instead of `String.split/1` in other places if you'd like. I really just don't want code examples out there that use it to split command arguments due to them not handling double quotes.

Another option is to use `System.shell/2` instead of `System.cmd/3`/`MuonTrap.cmd/3`. I have a preference for splitting out the arguments in Elixir rather than spawning a shell to do it for me, but I do find `System.shell/2` convenient.

---

If I understand correctly, it's `sh` that handles the double quotes.
So if we want to handle the double quotes, we can just use `:os.cmd` or `System.shell` as you said, right?

---

Right.

---

Thanks. I changed the commit and also changed the PR title.

How about b465c3acfb82bba2583bfbfe90d66172af6f14bc?

---

Looks good. Thanks!